### ISSUE TITLE
Unique vendor folder / Unique dependencies

### ISSUE THREAD

**Issue by celorodovalho** [10quality/wpmvc#16]
Hi!

I recently discovered this plugin, before I used to develop my plugins using tombener/wp-mvc, but seems it is abandoned and also restrict due some security issues.

I was reading the Documentation and tutorials and playing with some test plugins, so I faced a funny fact:

- There is one vendor dependencies for each plugin/theme?
- If I decide to develop 10 plugins using this repo, so I'm going to have 10 copies of the same dependencies?

There's no way to unify this? I mean, have a single dependency's vendor folder with all the reusable libraries?

So far I could understand, we can use Wordpress with composer, I'm personally using it on my personal projects/themes/plugins, and I used to have the following structure:
![image](https://user-images.githubusercontent.com/475722/97751150-2465ef00-1af2-11eb-9555-1454bd6055e1.png)

So, maybe you guys can advise how to use 10quality/wpmvc without to duplicate this dependencies, because I really couldn't understand.

**Comment by alexkuc** [comment]
While looking for some MVC and WP stuff, I have come across your post.

I think your question is not strictly related to this repository but rather a more generic one - "How can I share `vendor` across multiple projects?"

Here are some links which might help:

- https://laracasts.com/discuss/channels/laravel/using-the-same-vendor-directory-for-multiple-projects
- https://stackoverflow.com/questions/30000063/one-shared-vendor-with-two-projects

TL;DR: unless you are going to use *exactly* the same dependencies across those projects, it is regarded as a bad design choice; if you do use exactly the same dependencies, you can simply set up a symlink and share `vendor` dir

**Comment by str** [comment]
The plugins and themes you develop do not know how WordPress is installed, so there is now way to be sure WordPress is using composer or not. 

Let's imagine we have 2 plugins developed by 2 different providers, once uses another-provider/dependency-lib v1.2 and the other one uses the same dependency v1.3,  If y the v1.2 loads first, the v1.3 will not load,.The plugins do not load isolated, so if the autoloader already loaded a version, it will not load it again. The versions conflict can't be avoided, so there is no need to use the exactly same dependencies.

Having said that, I do think it is a good idea to have a shared folder for all vendor files. I thin the wp-content folder could be a good place BUT it's also tricky because [the path needs to be defined in the composer.json file,](https://stackoverflow.com/a/29458661/613408) and WordPress could be set to have WP_CONTENT_DIR pointing to another folder different that the standard folder.

**Comment by alexkuc** [comment]
Actually, after considering the OP's question, I am leaning towards the idea that OP (@celorodovalho) is looking for something like a monorepo (quite popular in Node ecosystem). Afterall, I think it is sensible to assume that if you want to control dependencies from a single point, you are going to VCS them under a "single roof". However, if it is a question of effective storage usage, then it is more appropriate to consider deduplication instead (i.e. address the issue from storage level).

**Comment by amostajo** [comment]
@celorodovalho @str @alexkuc The dependencies are shared between plugins and themes using the framework thanks to composer.

There is no need to create shared dependencies, as composer is intelligent enough to only load one source of each file.

Load priority will be handled by WordPress, meaning that wich ever plugin or theme was loaded first, those are the dependencies that will be used.

So for example. if "**Plugin A"** tries to use the `Bridge` class from the `wpmvc-core` dependency, then WordPress loads "**Plugin B**", it will utilize the class loaded from "**Plugin A**" dependencies.

------------
You can turn on the project and load dependencies from a WordPress root level as well, although the project will need some changes.
